File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    private List<List<String>> res;
    private List<String> list;
    private boolean isPalindrome(String s){
        int start = 0, end = s.length() - 1;
        while (start <= end){
            if(s.charAt(start) != s.charAt(end)){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
    
    private void generatePalindromePartition(String str) {
        if(str.isEmpty()){
            res.add(new ArrayList<>(list));
            return;
        }
        
        for (int i = 0; i < str.length(); i++) {
            String prefix = str.substring(0, i + 1);
            if(isPalindrome(prefix)){
                list.add(prefix);
                generatePalindromePartition(str.substring(i + 1));
                list.removeLast();
            }
        }
    }
    
    public List<List<String>> partition(String s) {
        res = new ArrayList<>();
        list = new ArrayList<>();
        generatePalindromePartition(s);
        return res;
    }
}

File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    bool isPallindrome(string &s, int start, int end)
    {
        while(start <= end)
        {
            if(s[start] != s[end])
            return false;

            start++;
            end--;
        }
        return true;
    }

    vector<string> temp;
    vector<vector<string>> res;
    void backTrack(string &s, int index)
    {
        if(index == s.length())
        {
            res.push_back(temp);
            return;
        }

        string curr = "";
        for(int i = index; i < s.length(); i++)
        {
            // generating the string
            curr.push_back(s[i]);

            if(isPallindrome(s, index, i))
            {
                temp.push_back(curr);
                backTrack(s, i + 1);

                // backTracking
                temp.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) 
    {
        backTrack(s, 0);
        return res;
    }
};